GPU/GPUTracking/TPCClusterFinder/GPUTPCClusterFinder.cxx

```diff
@@ -68,7 +68,7 @@ void GPUTPCClusterFinder::RegisterMemoryAllocation()
 void GPUTPCClusterFinder::SetMaxData(const GPUTrackingInOutPointers& io)
 {
   mNMaxPeaks = 0.5f * mNMaxDigits;
-  mNMaxClusters = 0.2f * mNMaxPeaks;
+  mNMaxClusters = mNMaxPeaks;
```

Comment:
We should be careful here not to blow up this value too much, because this will lead to excessive memory usage.
Number of clusters should at least be < 0.5*nDigits I guess.
Perhaps the scaling factor is not the problem, but we have to add an offset?

Second, I have started to collect these scaling factors in a common file.
Could you replace this by mNMaxClusters = mRec->MemoryScalers()->NTPCClusters(mNMaxDigits).
This scaler also already contains an offset. (See GPUTracking/Base/GPUMemorySizeScalers.h)

The buffers for the steps down the chain use this mechanism, so we should unify this.
Perhaps, could you run few simulations and estimate what would be a reasonable factor here.
And I'd rather have a large offset than a large scaling factor to compensate for fluctuations.

Reply:
Finally got back to this. Here are some numbers i got from  simulations:
- Number of peaks is at < 0.2 * number of digits (something like 0.17 but this seems to grow slowly with the number of events)
- There are about 10% fewer cluster than peaks. The noise suppression doesn't remove that many cluster
My suggestion would be to use the memory scaler for `mNMaxPeaks`, but keep `mNMaxClusters = mNMaxPeaks`.

Another place where we could save memory are the cluster buckets. Highest occupancy i saw was at 13% of the bucket size! Just replacing `mNMaxDigits` with `mNMaxPeaks` for calculating the max bucket size might do the trick?

Reply:
It makes sense that nPeaks / nDigits grows with occupancy. So we should put an upper bound here.
What makes me wonder a bit: with your patch we have nMaxClusters = 0.2 * nDigits, while before we had 0.1 * nDigits, or do I miss something? And before we were not running out of memory, at least not for large events. I think we should have the scalers tuned for high occupancy (Pb-Pb in time frame), and handle any fluctuations with an offset. Are you sure we need the 0.2 factor?

Reply:
I believe we need a factor of > 0.1. The number of cluster i saw was at something like 0.16 * nDigits.  (Maybe bring this factor down by retuning the noise suppression?) 

So nMaxClusters was too small the entire time. Why was this never problem?
The answer is: nMaxClusters is never used to allocate any memory...
There are two places where it could be used:
-  For allocating the buffer of filtered peaks, [GPUTPCClusterFinder.cxx:52](https://github.com/AliceO2Group/AliceO2/blob/5abb30e498fea88335cf79225c1a46965d90973f/GPU/GPUTracking/TPCClusterFinder/GPUTPCClusterFinder.cxx#L52):
` computePointerWithAlignment(mem, mPfilteredPeaks, mNMaxPeaks);`
- and the bucket size, [GPUTPCClusterFinder.cxx:72](https://github.com/AliceO2Group/AliceO2/blob/5abb30e498fea88335cf79225c1a46965d90973f/GPU/GPUTracking/TPCClusterFinder/GPUTPCClusterFinder.cxx#L72):
`  mNMaxClusterPerRow = 0.01f * mNMaxDigits;`

Until now all of these buffers have been way too big. Both peak buffer had a size of 0.5 * nDigits, when < 0.2 * nDigits is probably sufficient. And the bucket size maxClusterPerRow was set to 0.01 * nDigits, when 0.01 * nPeaks seems sufficient.
Now this makes me wonder where the cluster finder even overflowed in the first place?

Reply:
@fweig : OK, thx, that explains everything. The overflow was with some noisy real data, that was very "uneven", and there was not much data at all. So the scaling did not cover the fluctuations, and the offsets were too small.
I think we should be fine now. I'll merge this as soon as it passed the CI.
If we can use mNMaxClusters to reduce some memory for the filtered peaks and the bucket size, that makes sense, and please go ahead and do this in a separate PR.